### PR TITLE
Build and publish build image.

### LINKED ISSUE #5296: `loki-build-image` should be built and published via CI
**Is your feature request related to a problem? Please describe.**
Following https://github.com/grafana/loki/issues/5294, which raised a flag about the build image not being published after https://github.com/grafana/loki/pull/5250 was merged, I discovered we are publishing this image from our workstations.

**Describe the solution you'd like**
We should build and publish this image via CI.

**Describe alternatives you've considered**
Pushing from our workstations does work, but this can introduce inconsistencies or leak local changes into the image which are not pushed. We also currently rely on tribal knowledge and memory to remember to push a new build image when necessary.

**Additional context**
The image is pushed with `make build-image-push`.

### PR DESCRIPTION
**Pull request by jeschkies**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #5296

**Special notes for your reviewer**:

**Checklist**
- [x] Documentation added
- [ ] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

File: .drone/drone.jsonnet [hunk 1]
```diff
@@ -310,6 +310,18 @@ local manifest(apps) = pipeline('manifest') {
...
 [
+  pipeline('build-image') {
+    workspace: {
+      base: '/src',
+      path: 'loki',
+    },
+    steps: [
+      make('build-image-push', container=false) {
+        depends_on: ['clone'],
+        when: condition('include').tagMain,
```
[the remark this answers is not in the document]
Comment (dannykopping) [reply .drone/drone.jsonnet:321]: Here's one I pinched from elsewhere:

```jsonnet

  OnModifiedPaths(paths):: {
    trigger+: {
      paths+: {
        include:
          std.set(
            (if 'include' in super then super.include else [])
            + paths
          ),
        exclude: ['somebogusfilename'],  // see https://github.com/meltwater/drone-convert-pathschanged/issues/26
      },
    },
  },
```

File: .drone/drone.jsonnet [hunk 2]
```diff
@@ -310,6 +315,25 @@ local manifest(apps) = pipeline('manifest') {
...
+        depends_on: ['clone'],
+        commands: [
+          'apk add --no-cache bash make',
+          'make BUILD_IN_CONTAINER=false build-image'
+        ],
+        when: condition('include').path('loki-build-image/**'),
+        // when: condition('include').tagMain + condition('include').path('loki-build-image/**'),
+      },
+    ],
+  },
```
Comment (jeschkies) [inline .drone/drone.jsonnet:336]: I'm not sure what the best way is to build the image. `make loki-build-image` depends on `img`. However, for the other images we use the Docker plugin for Drone CI. @dannykopping what would you use?

Comment (dannykopping) [reply .drone/drone.jsonnet:336]: I think you should chat with the original author @slim-bean; I have no idea

File: docs/sources/maintaining/release-loki-build-image.md [hunk 3]
```diff
@@ -5,12 +5,15 @@ title: Releasing Loki Build Image
 
 The [`loki-build-image`](https://github.com/grafana/loki/tree/master/loki-build-image) is the Docker image used to run tests and build Grafana Loki binaries in CI.
 
-## How To Perform a Release
+The image is released with any change to `./loki-build-image` on `main`.
 
-1. Update `BUILD_IMAGE_VERSION` in the `Makefile`
+## How To Use a New Release
+
+1. Update the version tag of the `loki-build-image` pipeline.
```
Comment (dannykopping) [inline docs/sources/maintaining/release-loki-build-image.md:12]:
```suggestion
1. Update the version tag of the `loki-build-image` pipeline defined in `.drone/drone.jsonnet` (search for `pipeline('loki-build-image')`).
```

Alos: how does one retrieve this version tag?

**Review by dannykopping — APPROVED** [review]
LGTM, added a suggestion and a question.

Looks like you found a way around your chicken-and-egg problem, nice!

**Comment by jeschkies** [comment]
> Looks like you found a way around your chicken-and-egg problem, nice!

It will require two PRs but I think that's fine.

**Pull request merged**